crates/apollo-compiler/src/validation/variable.rs

```diff
@@ -13,81 +16,129 @@ pub fn validate_variable_definitions(
 ) -> Vec<ApolloDiagnostic> {
     let mut diagnostics = Vec::new();
 
+    let mut seen: HashMap<&str, &VariableDefinition> = HashMap::new();
     for variable in variables.iter() {
         diagnostics.extend(db.validate_directives(
             variable.directives().to_vec(),
             hir::DirectiveLocation::VariableDefinition,
         ));
+
+        let ty = variable.ty();
+        if !ty.is_input_type(db.upcast()) {
+            let type_def = ty.type_def(db.upcast());
+            if let Some(type_def) = type_def {
+                let ty_name = type_def.kind();
+                diagnostics.push(
+                    ApolloDiagnostic::new(db, variable.loc().into(), DiagnosticData::InputType {
+                        name: variable.name().into(),
+                        ty: ty_name,
+                    })
+                    .label(Label::new(ty.loc().unwrap(), format!("this is of `{ty_name}` type")))
+                    .help("objects, unions, and interfaces cannot be used as variables can only be of input type"),
```

Comment:
kinda nitpicky but the first time I read this as "objects, unions, and interfaces cannot be used as variables", with a different meaning of _as_ 😅 
```suggestion
                    .help("objects, unions, and interfaces cannot be used because variables can only be of input type"),
```

Reply:
hah fair fair